packages/ui/readme.md

```diff
@@ -1,28 +1,230 @@
+  - HeaderWithDropdown/
+    - index.tsx
+    - Dropdown.tsx ✅ Correct, if Dropdown is only ever used inside HeaderWithDropdown
+```
+
+```
+- src/components/
+  - Dropdown.tsx ❌ Wrong, if Dropdown is only ever used inside HeaderWithDropdown
+  - HeaderWithDropdown/
+    - index.tsx
+```
+
+(One exception to this rule: if you're developing a component that will eventually be used by multiple other components, and just happens to be a child of only a single component at the moment, you can leave it as a sibling of its parent.)
+
+### Components should be located at the most specific possible directory level.
+
+For example, if the `Dropdown` component is used by both `HeaderWithDropdown` and `Menu` components, `Dropdown` should be placed in the lowest-level directory that contains both `HeaderWithDropdown` and `Menu`:
+
+```
+- src/components/
+  - SomeCommonParentOfBothHeaderWithDropdownAndMenu/
+    - index.tsx
+    - HeaderWithDropdown/
+      - index.tsx
+      - index.test.tsx
+    - Menu/
+      - index.tsx
+      - index.test.tsx
+    - Dropdown.tsx ✅ Correct - Dropdown is used by both Menu and HeaderWithDropdown, so it's a sibling to both
+```
+
+This, as opposed to e.g., placing it inside the `HeaderWithDropdown` directory (and then importing it from there in `Menu`), or inside a root-level directory. This way, components are nested as closely to the components using them as possible.
+
+```
+- src/components/
+  - SomeCommonParentOfBothHeaderWithDropdownAndMenu/
+    - index.tsx
+    - HeaderWithDropdown/
+      - index.tsx
+      - index.test.tsx
+      - Dropdown.tsx ❌ Wrong - Menu shouldn't be importing a child of HeaderWithDropdown
+    - Menu/
+      - index.tsx
+      - index.test.tsx
+```
+
+### Components must be the default export of the component file.
+
+This allows them to be imported as the default, rather than destructuring an object. It also encourages UI library maintainers to only export one component per file.
+
+### Component props should be typed inline in the component function definition.
+
+For example:
+
+```tsx
+export default function MyComponent({ color }: { color: Color }) {
+  // ...
+}
+```
+
+This makes component code easy to read. An exception to this may be made when props are especially complex, such as when they are conditional.
+
+There is no need to export the props interface, since consumers can simply use `React.ComponentProps<typeof ComponentName>` to extract the props interface from the component.
+
+### Components must not accept `className` or `style` props.
+
+This ensures that each component is internally responsible for its own styling.
+
+Any variations of the component's appearance must be controlled via props like `variant`, `state`, etc. — not by kitchen-sink props like `className` and `style`, which allow arbitrary changes to be made that could interfere with the internal structure of the component and cause visual inconsistencies between instances of Penumbra UI components.
```

Comment:
question: even though I agree on this idea, there are cases where `className` existence is important. For example, take a look at the [AssetsTable Row implementation in minifront](https://github.com/penumbra-zone/web/blob/main/apps/minifront/src/components/dashboard/assets-table/index.tsx#L79). When the TableRow with class 'group' is hovered, the inner Link changes its opacity. How can we handle such case?

Reply:
Oh man, good find of an edge case... Hm.

Went down a rabbit hole trying to solve this problem, and here's what I came up with:

```tsx
            <TableBody>
              {account.balances.map((assetBalance, index) => (
                <TableRow key={index}>
                  <TableCell>
                    <ValueViewComponent view={assetBalance.balanceView} />
                  </TableCell>
                  <TableCell>
                    <EquivalentValues valueView={assetBalance.balanceView} />
                  </TableCell>
                  <TableCell>
                    <Link
                      className='transition md:opacity-0 [tr:hover>td>&]:opacity-100'
                      to={getTradeLink(assetBalance)}
                    >
                      <Button variant='secondary' size='md'>
                        Trade
                      </Button>
                    </Link>
                  </TableCell>
                </TableRow>
              ))}
            </TableBody>
```

I've removed `group` entirely, and used nesting selectors to get the same effect. (I tried using `group` in combination with `&` ([docs](https://tailwindcss.com/docs/hover-focus-and-other-states#arbitrary-groups)) to get the same effect, but it didn't work. I suspect I'm doing something wrong, and could fix this with a bit more work, but it didn't seem worth the time right now.)

This isn't a great solution, so I'll keep playing with it once I cross this particular bridge. But for now, I think we can still avoid `className`/`style` props.

Reply:
Wow, didn't know it's possible with Tailwind! Thanks for showing this. It probably worth being documented